src/type.js

```diff
@@ -0,0 +1,249 @@
+'use strict';
+
+import { core } from 'metal';
+
+const validators = {
+	any: () => true,
+	array: validateType('array'),
+	bool: validateType('boolean'),
+	func: validateType('function'),
+	number: validateType('number'),
+	object: validateType('object'),
+	string: validateType('string'),
+
+	arrayOf: (validator) => {
+		return composeValidator(
+			arrayOfValidatorFactory,
+			validator
+		);
+	},
+	instanceOf: (expectedClass) => {
+		return composeValidator(
+			instanceOfValidatorFactory,
+			expectedClass
+		);
+	},
+	objectOf: (validator) => {
+		return composeValidator(
+			objectOfValidatorFactory,
+			validator
+		);
+	},
+	oneOfType: (arrayOfTypeValidators) => {
+		return composeValidator(
+			oneOfTypeValidatorFactory,
+			arrayOfTypeValidators
+		);
+	},
+	shape: (shape) => {
+		return composeValidator(
+			shapeValidatorFactory,
+			shape
+		);
+	}
+};
+
+/**
+ * Creates a validator that checks the values of an array against a type.
+ * @param {!function} validator Type validator to check each index against.
+ * @param {boolean} noWarn Determines whether warnings should be emitted.
+ * @return {function} Validator.
+ */
+function arrayOfValidatorFactory(validator, noWarn) {
+	return (value, name, context) => {
+		if (!Array.isArray(value)) {
+			return warning('Expected an array.', name, context, noWarn);
+		} else {
+			const testArray = value.every(
+				item => {
+					return validator.noWarn(item, name);
```

Comment:
I like this code much better :D

But if we could avoid creating two functions we could also get rid of `composeValidator` (you mentioned that it could allow us to implement a `isRequired`, but I think that would be best as a State config, specially since the validator won't run if the value isn't set anyway, so it wouldn't be able to assert that it's required). Removing it will make the code even clearer.

One way of removing it would be to pass the `noWarn` flag as the fourth parameter for the validator. Another is to store a flag to stop warning in a variable of this file, outside the function, setting it to `true` before the loop and to previous value again at the end, like this:

```javascript
var _noWarn  = false;
function arrayOfValidatorFactory(validator) {
  // ...
  var prevNoWarn = _noWarn;
  _noWarn = true;
  const testArray = value.every(item => validator(item, name));
  _noWarn = prevNoWarn;
  // ...
}
```

Another option I just thought about is to have `metal-state` itself warn if the validator returns a string. That way people can easily add warnings to their own validators too. Then you don't need to worry about the nested validators warning more than once. Also, the logic here to suppress the warnings on production environment could just have validators return true instead of the string and all would work well. What do you think? If you like the idea, could you send a pr to `metal-state` adding the warnings for string validator returns? :)